File: Tutorials/C++ tutorials/Tutorial 2 - aircraft (DA40-NG)/Source files/tutorial_aircraft/DA40-NG.cpp
#include "DA40-NG.h"

// Definition of bone static members
int bones::propeller;
int bones::wheel_front;
int bones::wheel_right;
int bones::wheel_left;
int bones::elevator_right;
int bones::elevator_left;
int bones::rudder;
int bones::aileron_right;
int bones::aileron_left;
int bones::rudder_pedal_right;
int bones::rudder_pedal_left;
int bones::brake_pedal_left;
int bones::brake_pedal_right;
int bones::throttle_handle;

// Definition of mesh static members
int meshes::prop_blur;
int meshes::blade_one;
int meshes::blade_two;
int meshes::blade_three;

// Definition of sound static members
int sounds::rumble;
int sounds::eng_exterior;
int sounds::eng_interior;

// Definition of sound source static members
int sources::exterior;
int sources::interior;
int sources::rumble;

// Register the DA40 class as a client
namespace
{
	IFC_REGISTER_CLIENT(DA40);
};

//For now, it is also needed to register our mod, for that you have to create class which is derived from ot::script_module.
class register_mod : public ot::script_module
{
};

namespace dummy
{
	IFC_REGISTER_CLIENT(register_mod);
};

// Utility function to clamp a value within a specified range
float DA40::clamp(float val, float minval, float maxval)
{
	if (val < minval)
	{
		return minval;
	}
	else if (val > maxval)
	{
		return maxval;
	}
	else
	{
		return val;
	}
}

// Function to handle events on engine start/stop
void DA40::engine(int flags, uint code, uint channel, int handler_id)
{
	//safety checks for jsbsim pointer to avoid crashes
	if (!this->jsbsim) 
	{
		return;
	}

	// Toggle the engine state
	this->started ^= 1;

	// Start or stop the engine based on the state
	if (this->started == 1)
	{
		this->jsbsim->operator()("propulsion/starter_cmd", 1);
		this->jsbsim->operator()("propulsion/magneto_cmd", 1);
	}
	else
	{
		this->jsbsim->operator()("propulsion/starter_cmd", 0);
		this->jsbsim->operator()("propulsion/magneto_cmd", 0);
	}
}

// Function to handle landing lights events
void DA40::landing_lights(float val, uint code, uint channel, int handler_id)
{
	// Toggle landing lights based on the input value
	light_mask(0x3, val > 0);
};

// Function to handle navigation lights events
void DA40::navigation_lights(float val, uint code, uint channel, int handler_id)
{
	// Toggle navigation lights based on the input value
	light_mask(0x3, val > 0, nav_light_offset);
};

// Initialize chassis parameters for the DA40
ot::chassis_params DA40::init_chassis(const coid::charstr& params)
{
	// Define bone members
	bones::propeller = get_joint_id("propel");
	bones::wheel_front = get_joint_id("front_wheel");
	bones::wheel_right = get_joint_id("right_wheel");
	bones::wheel_left = get_joint_id("left_wheel");
	bones::elevator_right = get_joint_id("elevator_right");
	bones::elevator_left = get_joint_id("elevator_left");
	bones::rudder = get_joint_id("rudder");
	bones::aileron_right = get_joint_id("flap_right");
	bones::aileron_left = get_joint_id("flap_left");
	bones::rudder_pedal_right = get_joint_id("rudder_pedal_right");
	bones::rudder_pedal_left = get_joint_id("rudder_pedal_left");
	bones::brake_pedal_left = get_joint_id("brake_pedal_left");
	bones::brake_pedal_right = get_joint_id("brake_pedal_right");
	bones::throttle_handle = get_joint_id("throttle_lever");

	// Define mesh members 
	meshes::prop_blur = get_mesh_id("propel_blur");
	meshes::blade_one = get_mesh_id("main_blade_01");
	meshes::blade_two = get_mesh_id("main_blade_02");
	meshes::blade_three = get_mesh_id("main_blade_03");

	// Define sound members 
	sounds::rumble = load_sound("sounds/engine/engn1.ogg");
	sounds::eng_exterior = load_sound("sounds/engine/engn1_out.ogg");
	sounds::eng_interior = load_sound("sounds/engine/engn1_inn.ogg");
	
	// Define sound source members 
	sources::exterior = add_sound_emitter_id(bones::propeller, 1, 5.0f);
	sources::interior = add_sound_emitter_id(bones::propeller, -1, 1.2f);
	sources::rumble = add_sound_emitter_id(bones::propeller, 0, 2.0f);

	// Set up landing lights
	light_params = { .size = 0.1f, .angle = 100.f, .edge = 0.25f, .intensity = 5.f, .color = { 1.0f, 1.0f, 1.0f, 0.0f }, .fadeout = 0.05f,  };
	add_spot_light({ 4.5f, 1.08f, 0.98f }, {-0.1f, 1.f, 0.3f}, light_params);
	add_spot_light({ -4.5f, 1.08f, 0.98f }, {0.1f, 1.f, 0.3f}, light_params);

	// Set up navigation lights
	light_params = { .size = 0.035f, .angle = 100.f, .edge = 1.f, .intensity = 20.f, .color = { 1.0f, 1.0f, 1.0f, 0.0f }, .range = 0.0001f, .fadeout = 0.1f, };
	nav_light_offset = add_point_light({ 5.08f, 0.18f, 1.33f }, light_params);

	light_params.color = { 0.f, 1.f, 0.f, 0.f };
	add_point_light({-5.05f, 0.18f, 1.33f}, light_params);

	// Register handlers
	register_axis_handler("air/lights/landing_lights", &DA40::landing_lights, { .minval = 0.f, .maxval = 1.f, .cenvel = 0.f, .vel = 10.f });
	register_axis_handler("air/lights/nav_lights", &DA40::navigation_lights, { .minval = 0.f, .maxval = 1.f, .cenvel = 0.f, .vel = 10.f });
	register_event_handler("air/engines/on", &DA40::engine);

	// Return chassis parameters
	return {
		.mass = 1310,
		.com_offset = {0.0f, 0.0f, 0.2f},
	};
}

// Initialize the DA40 aircraft
void DA40::initialize(bool reload)
{
	//Get JSBSim interface
	this->jsbsim = jsb();
	//Get geomob interface
	this->geom = get_geomob(0);
	//Get sound interface
	this->snd = sound();

	// Set initial fps position
	set_fps_camera_pos({ 0.f, 1.f, 1.4f });

	// Set initial values for the aircraft
	this->started = false;

	//safety check for jsbsim pointer to avoid crashes
	if (!this->jsbsim)
	{
		return;
	}

	// Set initial JSBSim properties
	this->jsbsim->operator()("propulsion/starter_cmd", 0);
	this->jsbsim->operator()("fcs/elevator-pos-rad", 0);
	this->jsbsim->operator()("fcs/rudder-pos-rad", 0);
	this->jsbsim->operator()("fcs/right-aileron-pos-rad", 0);
	this->jsbsim->operator()("fcs/left-aileron-pos-rad", 0);
	this->jsbsim->operator()("fcs/right-brake-cmd-norm", 0);
	this->jsbsim->operator()("fcs/left-brake-cmd-norm", 0);
	this->jsbsim->operator()("fcs/throttle-cmd-norm[0]", 0);
	this->jsbsim->operator()("fcs/mixture-cmd-norm[0]", 0);
}

//Update model instance each frame
void DA40::update_frame(float dt)
{
	//safety check for pointers to avoid crashes
	if (!this->jsbsim || !this->geom || !this->snd)
	{
		return;
	}

	//Store values based on JSBSim data
	float eng_rpm = static_cast<float>(this->jsbsim->operator()("propulsion/engine[0]/engine-rpm"));
	float wheel_speed = static_cast<float>(this->jsbsim->operator()("gear/unit[0]/wheel-speed-fps"));
	float elev_pos_rad = static_cast<float>(this->jsbsim->operator()("fcs/elevator-pos-rad"));

	// Update the orientation of different components
	this->geom->rotate_joint(bones::propeller, dt * (2 * static_cast<float>(PI)) * eng_rpm, { 0.f, 1.f, 0.f });
	this->geom->rotate_joint(bones::wheel_front, dt * static_cast<float>(PI) * (wheel_speed / 5), { -1.f, 0.f, 0.f });
	this->geom->rotate_joint(bones::wheel_right, dt * static_cast<float>(PI) * (wheel_speed / 5), { -1.f, 0.f, 0.f });
	this->geom->rotate_joint(bones::wheel_left, dt * static_cast<float>(PI) * (wheel_speed / 5), { -1.f, 0.f, 0.f });
	this->geom->rotate_joint(bones::elevator_right, elev_pos_rad, { 1.f, 0.f, 0.f });
	this->geom->rotate_joint(bones::elevator_left, elev_pos_rad, { 1.f, 0.f, 0.f });

	this->geom->rotate_joint_orig(bones::rudder, static_cast<float>(this->jsbsim->operator()("fcs/rudder-pos-rad")), { 0.f, 0.f, -1.f });
	this->geom->rotate_joint_orig(bones::aileron_right, static_cast<float>(this->jsbsim->operator()("fcs/right-aileron-pos-rad")), { -1.f, 0.f, 0.f });
	this->geom->rotate_joint_orig(bones::aileron_left, static_cast<float>(this->jsbsim->operator()("fcs/left-aileron-pos-rad")), { 1.f, 0.f, 0.f });
	this->geom->rotate_joint_orig(bones::brake_pedal_left, static_cast<float>(this->jsbsim->operator()("fcs/left-brake-cmd-norm")) * 0.7f, { -1.f, 0.f, 0.f });
	this->geom->rotate_joint_orig(bones::brake_pedal_right, static_cast<float>(this->jsbsim->operator()("fcs/right-brake-cmd-norm")) * 0.7f, { -1.f, 0.f, 0.f });
   
	if (static_cast<float>(this->jsbsim->operator()("fcs/rudder-pos-rad")) > 0)
	{
		this->geom->move_joint_orig(bones::rudder_pedal_left, { 0, static_cast<float>(this->jsbsim->operator()("fcs/rudder-pos-rad")) / 5, 0 });
	}
	else if (static_cast<float>(this->jsbsim->operator()("fcs/rudder-pos-rad")) < 0)
	{
		this->geom->move_joint_orig(bones::rudder_pedal_right, { 0, - static_cast<float>(this->jsbsim->operator()("fcs/rudder-pos-rad")) / 5,0 });
	}

	this->geom->move_joint_orig(bones::throttle_handle, { 0, static_cast<float>(this->jsbsim->operator()("fcs/throttle-cmd-norm[0]")) / 7, 0 });

	// Update the visibility of propeller meshes, based on the engine rpm
	this->geom->set_mesh_visible_id(meshes::prop_blur, eng_rpm > 200.f);
	this->geom->set_mesh_visible_id(meshes::blade_one, eng_rpm < 300.f);
	this->geom->set_mesh_visible_id(meshes::blade_two, eng_rpm < 300.f);
	this->geom->set_mesh_visible_id(meshes::blade_three, eng_rpm < 300.f);

	// When engine runs, control sounds, based on the camera position & sound effects, based on engine RPM
	if (eng_rpm > 0)
	{
		if (get_camera_mode() == 0)
		{
			this->snd->stop(sources::exterior);

			this->snd->set_pitch(sources::interior, clamp(eng_rpm / 4000.f, 0.f, 0.6f));
			this->snd->set_gain(sources::interior, clamp(eng_rpm / 4000.f, 0.f, 0.6f));

			if (this->snd->is_playing(sources::interior))
			{
				this->snd->play_loop(sources::interior, sounds::eng_interior);
			}

			this->snd->set_pitch(sources::rumble, clamp(eng_rpm / 1200, 0.f, 2.f));
			this->snd->set_gain(sources::rumble, clamp(eng_rpm / 2000.f, 0.f, 1.5f));

			if (this->snd->is_playing(sources::rumble))
			{
				this->snd->play_loop(sources::rumble, sounds::rumble);
			}
		}
		else
		{
			this->snd->stop(sources::interior);

			this->snd->set_pitch(sources::exterior, clamp(eng_rpm / 4000.f, 0.f, 0.6f));
			this->snd->set_gain(sources::exterior, clamp(eng_rpm / 4000.f, 0.f, 0.6f));

			if (this->snd->is_playing(sources::exterior))
			{
				this->snd->play_loop(sources::exterior, sounds::eng_exterior);
			}

			this->snd->set_pitch(sources::rumble, clamp(eng_rpm / 1200, 0.f, 2.f));
			this->snd->set_gain(sources::rumble, clamp(eng_rpm / 1300.f, 0.f, 2.f));

			if (this->snd->is_playing(sources::rumble))
			{
				this->snd->play_loop(sources::rumble, sounds::rumble);
			}
		}
	} // Stop sounds when the engine is not running
	else
	{
		this->snd->stop(sources::exterior);
		this->snd->stop(sources::interior);
		this->snd->stop(sources::rumble);
	}
}

File: Tutorials/example_models_js/packages/tutorial_aircraft_js/DA40-NG.js
//Welcome to the second tutorial on JavaScript programming in Outerra.
//We recommend reviewing previous tutorial (tutorial_car) before proceeding, as well as the corresponding tutorial on the Outerra/anteworld GitHub page ( https://github.com/Outerra/anteworld/wiki/Tutorial-2-%E2%80%90-Aircraft-(javascript) ), as it contains detailed information.

//Define global variables
//Used for joint rotation
const PI = 3.14159265358979323846;

//Create objects and initialize it's properties with default values (-1 for debug purpose)
//Bones
let bones = {
    propeller : -1,
    wheel_front : -1,
    wheel_right : -1,
    wheel_left : -1,
    elevator_right : -1,
    elevator_left : -1,
    rudder : -1,
    aileron_right : -1,
    aileron_left : -1,
    rudder_pedal_left : -1,
    rudder_pedal_right : -1,
    brake_pedal_left : -1,
    brake_pedal_right : -1,
    throttle_handle : -1,
    flap_left : -1,
    flap_right : -1,  
};

//Meshes
let meshes = {
    prop_blur : -1,
	blade_one : -1,
	blade_two : -1,
	blade_three : -1,
};

//Sounds 
let sounds = {
    rumble : -1,
    eng_int : -1,
    eng_ext : -1,
    prop_int : -1,
    prop_ext : -1,
};

//Sound sources
let sources = {
    rumble_int : -1,
    rumble_ext : -1,
    eng_int : -1,
    eng_ext : -1,
    prop_int : -1,
    prop_ext : -1,
};

//Used to clamp values.
function clamp(val, min, max) 
{
	if(val < min) return min; 
	else if(val > max) return max; 
	else return val;
}

//Gets called when engine starts/stops
function engine()
{
    //toggle the value every time this function is called 
    this.started ^= 1;
    
    if(this.started === 1)
    {
        //For JSBSim aircraft to start, the starter and magneto needs to be activated
        //Turn on engine (0 - turn off, 1 - turn on)
        this.jsbsim['propulsion/starter_cmd'] = 1;
        //debug tip: when the magneto isn't activated, it results in decreased piston engine power, with the rpm not exceeding 1000
        this.jsbsim['propulsion/magneto_cmd'] = 1;
    }
    else
    {
        this.jsbsim['propulsion/starter_cmd'] = 0;
        this.jsbsim['propulsion/magneto_cmd'] = 0;
    }
}
//"this.jsbsim['property']" is used, to set/get property belonging to JSBSim interface. 
//For information on JSBSim, visit the JSBSim & Aeromatic page (https://github.com/Outerra/anteworld/wiki/JSBSim-&-Aeromatic).
//For a list of usable JSBSim properties, refer to the JSBSim properties page (https://github.com/Outerra/anteworld/wiki/JSBSim-properties).

//Invoked once, to define the chassis for all instances
function init_chassis()
{
	//Get joins/bones
	bones.propeller = this.get_joint_id('propel');
	bones.wheel_front = this.get_joint_id('front_wheel');
	bones.wheel_right = this.get_joint_id('right_wheel');
	bones.wheel_left = this.get_joint_id('left_wheel');
	bones.elevator_right = this.get_joint_id('elevator_right'); 
	bones.elevator_left = this.get_joint_id('elevator_left');
	bones.rudder = this.get_joint_id('rudder');
	bones.aileron_right = this.get_joint_id('aileron_right'); 
	bones.aileron_left = this.get_joint_id('aileron_left');
    bones.flap_left = this.get_joint_id('flap_left');
    bones.flap_right = this.get_joint_id('flap_right');
	bones.rudder_pedal_left = this.get_joint_id('rudder_pedal_left');
	bones.rudder_pedal_right = this.get_joint_id('rudder_pedal_right');
	bones.brake_pedal_left = this.get_joint_id('brake_pedal_left');
	bones.brake_pedal_right = this.get_joint_id('brake_pedal_right');
	bones.throttle_handle = this.get_joint_id('throttle_lever');
	
	//Get mesh, for that use function get_mesh_id
	// 1.parameter - mesh name
	meshes.prop_blur = this.get_mesh_id('propel_blur');
	meshes.blade_one = this.get_mesh_id('main_blade_01#0@0');
	meshes.blade_two = this.get_mesh_id('main_blade_02#0@0');
	meshes.blade_three = this.get_mesh_id('main_blade_03#0@0');

	//Add lights
	//Landing lights
	let light_params = {color:{x:1,y:1,z:1}, angle:100, size:0.1, edge:0.25, intensity:5, fadeout:0.05};
	this.add_spot_light({x:4.5, y:1.08, z:0.98}, {x:-0.1, y:1, z: 0.3}, light_params);
	this.add_spot_light({x:-4.5, y:1.08, z:0.98}, {x:0.1, y:1, z: 0.3}, light_params);

	//Navigation lights
	light_params = {color:{x:0,y:1,z:0}, size:0.035, edge:1, range:0.0001, intensity:20, fadeout:0.1};
	let nav_light_offset =
	//Right green navigation light
	this.add_point_light({x:5.08, y:0.18, z:1.33}, light_params);
	//Left red navigation light
	light_params.color = {x:1,y:0,z:0};
	this.add_point_light({x:-5.08, y:0.18, z:1.33}, light_params);
	

	//Load sounds
  	//engine rumble - in this case it's used for the outside sounds and also the inside
    sounds.rumble = this.load_sound("sounds/engine/engn1.ogg");
	
	//Interior sounds - will be heard from inside the plane
    //engine
	sounds.eng_ext = this.load_sound("sounds/engine/engn1_out.ogg");
    //propeller
	sounds.prop_ext = this.load_sound("sounds/engine/prop1_out.ogg");

	//Exterior sounds - will be heard from outside the plane 
    //engine
	sounds.eng_int = this.load_sound("sounds/engine/engn1_inn.ogg"); 
    //propeller
	sounds.prop_int = this.load_sound("sounds/engine/prop1_out.ogg");
	
	//Add sound emitters
    //Interior emitters
	//For interior rumbling sound
    sources.rumble_int = this.add_sound_emitter_id(bones.propeller, -1, 0.5);
    //For interior engine sounds
	sources.eng_int = this.add_sound_emitter_id(bones.propeller, -1, 0.5);
	//For interior propeller sounds
	sources.prop_int = this.add_sound_emitter_id(bones.propeller, -1, 0.5);
    
    //Exterior emitters
    //For exterior rumbling sound
    sources.rumble_ext = this.add_sound_emitter_id(bones.propeller, 1, 3);
	//For exterior engine sounds 
	sources.eng_ext = this.add_sound_emitter_id(bones.propeller, 1, 3);
	//For exterior propeller sounds 
	sources.prop_ext = this.add_sound_emitter_id(bones.propeller, 1, 3);
    
    //Note: it is possible, to use one emitter for interior and exterior, like in previous tutorial
    
    //Add action handlers
	this.register_axis("air/lights/landing_lights", {minval: 0, maxval: 1, vel: 10, center: 0 }, function(v) { this.light_mask( 0x3,v > 0); });
    this.register_axis("air/lights/nav_lights", {minval: 0, maxval: 1, vel: 10, center: 0 }, function(v) { this.light_mask( 0x3,v > 0, nav_light_offset); });
   	
    /*Note: by default, JSBSim actions are handled internally, allowing gameplay without direct user intervention. However, sometimes it is needed to 
    customize or fine-tune aircraft behavior, therefore in this case, the following handlers are handled in script.*/
    this.register_event("air/engines/on", engine);
    this.register_axis("air/controls/aileron", { minval: -1, maxval: 1, center: 0.5, vel: 0.5, positions: 0 }, function(v){
       	this.jsbsim['fcs/aileron-cmd-norm'] = v;
    });
	this.register_axis("air/controls/elevator", { minval: -1, maxval: 1, center: 0.5, vel: 0.5, positions: 0 }, function(v){
        this.jsbsim['fcs/elevator-cmd-norm'] = -v;
    });
    this.register_axis("air/controls/brake", { minval: 0}, function(v) { 
        this.braking = v;
    	this.jsbsim['fcs/center-brake-cmd-norm'] = v;
        this.jsbsim['fcs/left-brake-cmd-norm'] = v;
        this.jsbsim['fcs/right-brake-cmd-norm'] = v;
    });
  
	return {
		mass: 1310,
		com: {x: 0.0, y: 0.0, z: 0.2},
	};
}

//Function initialize() is used to define per-instance parameters (similarly how vehicles use init_vehicle())
function initialize()
{
	//Get geomob interface
	this.geom = this.get_geomob(0);
	
	//Get JSBSim interface, to be able to work with JSBSim properties
	this.jsbsim = this.jsb();
	
	//Get sound interface
	this.snd = this.sound();
	
	//Set first person camera position 
	this.set_fps_camera_pos({x:0, y:1, z:1.4});
    
    this.started = 0;
    this.braking = 0;
	
	//Set initial value for JSBSim properties
	//Turn off engine (commands are normalized)
	this.jsbsim['propulsion/starter_cmd'] = 0;
    //Turn off magneto
	this.jsbsim['propulsion/magneto_cmd'] = 0;
	//Set initial throttle value to 0 
	this.jsbsim['fcs/throttle-cmd-norm[0]'] = 0;
    //Set initial throttle value to 0 
    this.jsbsim['fcs/mixture-cmd-norm'] = 0;
	//Set initial right brake value to 0 
	this.jsbsim['fcs/right-brake-cmd-norm'] = 0;
	//Set initial left brake value to 0 
	this.jsbsim['fcs/left-brake-cmd-norm'] = 0;
    
    //Set default pitch value for sound emitters
    this.snd.set_pitch(sources.rumble_ext, 1);
    this.snd.set_pitch(sources.rumble_int, 1);
    this.snd.set_pitch(sources.eng_ext, 1);
    this.snd.set_pitch(sources.eng_int, 1);
    this.snd.set_pitch(sources.prop_ext, 1);
    this.snd.set_pitch(sources.prop_int, 1);
    
    //Set default gain value for sound emitters
    this.snd.set_gain(sources.rumble_ext, 1);
    this.snd.set_gain(sources.rumble_int, 1);
    this.snd.set_gain(sources.eng_ext, 1);
    this.snd.set_gain(sources.eng_int, 1);
    this.snd.set_gain(sources.prop_ext, 1);
    this.snd.set_gain(sources.prop_int, 1);
}

//Invoked each frame to handle the internal state of the object
function update_frame(dt)
{	
	//Get engine rpm from JSBSim
	let propeller_rpm = this.jsbsim['propulsion/engine[0]/propeller-rpm'];
	//Get wheel speed from JSBSim
	let wheel_speed = this.jsbsim['gear/unit[0]/wheel-speed-fps'];
	//Get elevator rotation in radians
	let elev_pos_rad = this.jsbsim['fcs/elevator-pos-rad'];
	
	//rotate_joint - incremental, bone rotates by given angle every time, this function is called (given angle is added to joint current orientation)
	this.geom.rotate_joint(bones.propeller, dt * (2 * PI) * (propeller_rpm), {y:1});
	this.geom.rotate_joint(bones.wheel_front, dt * PI * (wheel_speed / 5), {x:-1});
	this.geom.rotate_joint(bones.wheel_right, dt * PI * (wheel_speed / 5), {x:-1});
	this.geom.rotate_joint(bones.wheel_left, dt * PI * (wheel_speed / 5), {x:-1});
    //rotate_joint_orig - rotate to given angle (from default orientation)
    this.geom.rotate_joint_orig(bones.elevator_right, elev_pos_rad, {x:1});
	this.geom.rotate_joint_orig(bones.elevator_left, elev_pos_rad, {x:1});
	this.geom.rotate_joint_orig(bones.rudder, this.jsbsim['fcs/rudder-pos-rad'], {z:-1});
	this.geom.rotate_joint_orig(bones.aileron_right, this.jsbsim['fcs/right-aileron-pos-rad'], {x:-1});
	this.geom.rotate_joint_orig(bones.aileron_left, this.jsbsim['fcs/left-aileron-pos-rad'], {x:1});
	this.geom.rotate_joint_orig(bones.flap_left, this.jsbsim['fcs/flap-pos-rad'], {x:1,y:0,z:0});
    this.geom.rotate_joint_orig(bones.flap_right, this.jsbsim['fcs/flap-pos-rad'], {x:1,y:0,z:0});
    this.geom.rotate_joint_orig(bones.brake_pedal_left, this.jsbsim['fcs/left-brake-cmd-norm'] * 0.7, {x:-1});
	this.geom.rotate_joint_orig(bones.brake_pedal_right, this.jsbsim['fcs/right-brake-cmd-norm'] * 0.7, {x:-1});

	//Rudder has 2 pedals for turning, therefore depending on the rudder position value from JSBSim, the left or right pedal will move
	if(this.jsbsim['fcs/rudder-pos-rad'] > 0)
	{
		this.geom.move_joint_orig(bones.rudder_pedal_left, {y:(this.jsbsim['fcs/rudder-pos-rad']/5)} );
	}
	else if(this.jsbsim['fcs/rudder-pos-rad'] < 0)
	{
		this.geom.move_joint_orig(bones.rudder_pedal_right, {y:-(this.jsbsim['fcs/rudder-pos-rad']/5)} );
	}	
	
	//Move throttle handle, depending on the throttle command value 
	this.geom.move_joint_orig(bones.throttle_handle, {y:(this.jsbsim['fcs/throttle-cmd-norm[0]'] / 7)});
	
	//Note: in this case the pitch/roll handle rotates without the need of scripting, because it's binded with inputs in the model
	
	//Use set_mesh_visible_id() function, to set propeller blur mesh visible, when condition is true and invisible otherwise
	//1.param - mesh ID
	//2.param - condition
	this.geom.set_mesh_visible_id(meshes.prop_blur, propeller_rpm > 200.0);
	this.geom.set_mesh_visible_id(meshes.blade_one, propeller_rpm < 300.0);
	this.geom.set_mesh_visible_id(meshes.blade_two, propeller_rpm < 300.0);
	this.geom.set_mesh_visible_id(meshes.blade_three, propeller_rpm < 300.0);
	
    //When engine runs, control sounds, based on the camera position & sound effects, based on engine RPM
	if(propeller_rpm > 0)
	{
		//Interior
		if (this.get_camera_mode() === 0 ) 
		{	
			//Turn off the exterior emitters, when player camera is inside the plane
			this.snd.stop(sources.eng_ext); 
            this.snd.stop(sources.prop_ext);
            this.snd.stop(sources.rumble_ext);
            
			//Calculate and set pitch for interior engine emitter (clamped between 1 and 2)
			this.snd.set_pitch(sources.eng_int, clamp(1 + propeller_rpm/4000, 1, 2));
			//Calculate and set gain for interior engine emitter (clamped between 0 and 0.5)
			this.snd.set_gain(sources.eng_int, clamp(propeller_rpm/5000, 0, 0.5));
			
            //If there is no sound already playing on given emitter, play loop
			if(!this.snd.is_playing(sources.eng_int)) 
			{
				//Play interior engine sound in a loop
				this.snd.play_loop(sources.eng_int, sounds.eng_int);
			}
            
			//Set gain for interior propeller emitter
			this.snd.set_gain(sources.prop_int, clamp(propeller_rpm/7000, 0, 0.5));
			
			if(!this.snd.is_playing(sources.prop_int)) 
			{
				//Play interior propeller sound in a loop
				this.snd.play_loop(sources.prop_int, sounds.prop_int);
			}
			
			//Set gain for interior rumble emitter
			this.snd.set_gain (sources.rumble_int, clamp(propeller_rpm/6000, 0, 0.5));
			
            if(!this.snd.is_playing(sources.rumble_int)) 
			{
				//Play rumble sound in a loop on interior emitter
				this.snd.play_loop(sources.rumble_int, sounds.rumble);
			}
        }
		//Exterior
		else 
		{
			//Turn off the interior emitters, when camera is outside the plane
			this.snd.stop (sources.eng_int);
			this.snd.stop (sources.prop_int);
			this.snd.stop (sources.rumble_int);
            
			//Set pitch for exterior engine emitter
			this.snd.set_pitch(sources.eng_ext, clamp(1 + propeller_rpm/1000, 1, 3));
			//Set gain for exterior engine emitter
			this.snd.set_gain (sources.eng_ext, clamp(propeller_rpm/450, 0.05, 2));
			
            if(!this.snd.is_playing(sources.eng_ext)) 
			{
				//Play exterior engine sound in a loop
				this.snd.play_loop(sources.eng_ext, sounds.eng_ext);
			}
            
            //Set gain for exterior propeller emitter
			this.snd.set_gain (sources.prop_ext, clamp(propeller_rpm/900, 0, 2));
            
            if(!this.snd.is_playing(sources.prop_ext)) 
			{
				//Play exterior propeller sound in a loop
				this.snd.play_loop(sources.prop_ext, sounds.prop_ext);
			}
			
			//Set gain for exterior rumble emitter
			this.snd.set_gain (sources.rumble_ext, clamp(propeller_rpm/1200, 0, 2));
			
			if(!this.snd.is_playing(sources.rumble_ext)) 
			{
				//Play rumble sound in a loop on exterior emitter
				this.snd.play_loop(sources.rumble_ext, sounds.rumble);
			}
		}
	}
	else 
	{
		//Stop all sounds playing on given emitters, when engine is turned off and propeller stops rotating
		this.snd.stop(sources.eng_ext);
		this.snd.stop(sources.eng_int);
        this.snd.stop(sources.prop_ext);
		this.snd.stop(sources.prop_int);
		this.snd.stop(sources.rumble_ext);
		this.snd.stop(sources.rumble_int);
    }	
    
    //Parking brake when in idle state 
    if(!this.started && propeller_rpm < 5)
    {
        this.jsbsim['fcs/center-brake-cmd-norm'] = 1;
        this.jsbsim['fcs/left-brake-cmd-norm'] = 1;
        this.jsbsim['fcs/right-brake-cmd-norm'] = 1;
    }
    else if (this.braking < 0.1)
    {
        this.jsbsim['fcs/center-brake-cmd-norm'] = 0;
        this.jsbsim['fcs/left-brake-cmd-norm'] = 0;
        this.jsbsim['fcs/right-brake-cmd-norm'] = 0;
    } 
}

